gluten-core/src/main/scala/io/glutenproject/execution/ScanTransformerFactory.scala

```diff
@@ -0,0 +1,115 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.execution
+
+import io.glutenproject.execution.FilterHandler.{flattenCondition, getLeftFilters}
+import io.glutenproject.expression.ExpressionConverter
+import io.glutenproject.sql.shims.SparkShimLoader
+
+import org.apache.spark.sql.connector.read.Scan
+import org.apache.spark.sql.execution.{FileSourceScanExec, FilterExec}
+import org.apache.spark.sql.execution.datasources.v2.{BatchScanExec, FileScan}
+
+import scala.reflect.runtime.{universe => ru}
+
+object ScanTransformerFactory {
+
+  private val IcebergScanClassName = "org.apache.iceberg.spark.source.SparkBatchQueryScan"
+  private val IcebergTransformerClassName = "io.glutenproject.execution.IcebergScanTransformer"
+
+  def createFileSourceScanTransformer(
+      scanExec: FileSourceScanExec,
+      reuseSubquery: Boolean,
+      validation: Boolean = false): FileSourceScanTransformer = {
+    // TODO: Add delta match here
+    val newPartitionFilters = if (validation) {
+      scanExec.partitionFilters
+    } else {
+      ExpressionConverter.transformDynamicPruningExpr(scanExec.partitionFilters, reuseSubquery)
+    }
+    new FileSourceScanTransformer(
+      scanExec.relation,
+      scanExec.output,
+      scanExec.requiredSchema,
+      newPartitionFilters,
+      scanExec.optionalBucketSet,
+      scanExec.optionalNumCoalescedBuckets,
+      scanExec.dataFilters,
+      scanExec.tableIdentifier,
+      scanExec.disableBucketedScan
+    )
+  }
+
+  def createFileSourceScanTransformer(
+      scanExec: FileSourceScanExec,
+      reuseSubquery: Boolean,
+      filter: FilterExec): FileSourceScanTransformer = {
+    val leftFilters =
+      getLeftFilters(scanExec.dataFilters, flattenCondition(filter.condition))
+    // transform BroadcastExchangeExec to ColumnarBroadcastExchangeExec in partitionFilters
+    val newPartitionFilters =
+      ExpressionConverter.transformDynamicPruningExpr(scanExec.partitionFilters, reuseSubquery)
+    new FileSourceScanTransformer(
+      scanExec.relation,
+      scanExec.output,
+      scanExec.requiredSchema,
+      newPartitionFilters,
+      scanExec.optionalBucketSet,
+      scanExec.optionalNumCoalescedBuckets,
+      scanExec.dataFilters ++ leftFilters,
+      scanExec.tableIdentifier,
+      scanExec.disableBucketedScan
+    )
+  }
+
+  def createBatchScanTransformer(
+      batchScanExec: BatchScanExec,
+      reuseSubquery: Boolean,
+      validation: Boolean = false): BatchScanTransformer = {
+    val newPartitionFilters = if (validation) {
+      batchScanExec.runtimeFilters
+    } else {
+      ExpressionConverter.transformDynamicPruningExpr(batchScanExec.runtimeFilters, reuseSubquery)
+    }
+    val scan = batchScanExec.scan
+    scan match {
+      case _ if scan.getClass.getName == IcebergScanClassName =>
```

Comment:
IIUC, supporting a new `Scan` type (e.g. adding a non-Iceberg data source) will require modifying this `match` code, as well as the `supportedBatchScan()` method below.
Would it please be possible to allow other plugins to register themselves, so that adding a new format will not require changing `ScanTransformerFactory` code?

Reply:
@rz-vastdata We can achieve this using the Service Loader, similar to `DataSourceRegister` in Spark. However, the current `ScanTransformer` in Gluten extends Spark's `ScanExec` and requires a constructor with parameters. This makes it a bit difficult to handle with Service Loader. I will think about whether there is another way to use the Service Loader.